lte/gateway/python/magma/enodebd/state_machines/enb_acs.py

```diff
@@ -218,3 +221,9 @@ def is_enodeb_connected(self) -> bool:
     @abstractmethod
     def stop_state_machine(self) -> None:
         pass
+
+    def download_asap(self, url: str, user_name: str, password: str, target_file_name: str, file_size: int, md5: str) -> None:
+        """
+        Send a request to download&upgrade the eNodeB ASAP
+        """
+        pass
```

Comment:
## 1
This should probably be marked as `@abstractmethod`. While this has been implemented for QRTB, if this method is called on other eNodeB handlers within `enodebd` service, this method will fail silently.


## 2
Some additional documentation is warranted here regarding the arguments. Does the `url` refer specifically to a password protected download link? Must it always, or is it possible for it refer to a download URL that is not password protected? And in that case, what use is the `user_name` and `password` - what would the caller pass in? There are some opinionated decisions being made here about exactly what is hiding behind this URL, but it's not clear to me as a reviewer that every download URL will conform to what's being required here.

One possible alternate function signature could be:
```suggestion
    def download_asap(self, download_metadata: Any) -> None:
        """
        Send a request to download&upgrade the eNodeB ASAP
        """
        # QRTB handler would check the type of download_metadata, and handle appropriately
        # Other eNodeB handlers might expect downloads to be specified differently, so maybe it needs its own type
        pass
```
This would require `enodebd_cli.py` to be more complicated to compensate, of course, but this is just one option. Either way, I think this method signature can't remain the way it is.